File: src/qrs-core-js/src/services/clock.ts
/**
 * Time abstraction (IoC). The real clock is the default; tests and consumers can
 * inject a fixed clock to make behavior deterministic.
 */
export interface IClock {
  /** Current time as epoch seconds (UTC). */
  now(): number;
}

export class SystemClock implements IClock {
  now(): number {
    return Math.floor(Date.now() / 1000);
  }
}

File: src/qrs-core-python/src/qrs/clock.py
"""Clock abstraction (IoC for time)."""

from __future__ import annotations

import time
from typing import Protocol

__all__ = ["IClock", "SystemClock"]


class IClock(Protocol):
    """Provides the current Unix epoch time in seconds."""

    def now(self) -> int: ...


class SystemClock:
    """The default clock: wall-clock time."""

    def now(self) -> int:
        return int(time.time())